alf/utils/data_buffer.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Classes for storing data for sampling."""
+
+import tensorflow as tf
+
+from tf_agents.utils import common as tfa_common
+
+
+class DataBuffer(tf.Module):
+    """A simple circular buffer supporting random sampling.
+    """
+
+    def __init__(self, tensor_spec: tf.TensorSpec, capacity,
+                 name="DataBuffer"):
+        """Create a DataBuffer.
+
+        Args:
+            tensor_spec (TensorSpec): spec for the data item (without batch
+                dimension) to be stored.
+            capacity (int): capacity of the buffer.
+            name (str): name of the buffer
+        """
+        super().__init__()
+        self._capacity = capacity
+        shape = shape = [capacity] + tensor_spec.shape.as_list()
```

Comment:
shape = [capacity] + tensor_spec.shape.as_list()

Reply:
Why not use  `TFUniformReplayBuffer` instead ?

Reply:
TFUniformReplayBuffer requires batch_size at construction, and each add_batch() call has to add a batch with the exactly same batch_size given to the constructor.

Reply:
ok , i see